RPDE - first go

odscjames opened the pull request.

jarofgreen committed:
rpde: First go at end point

odscjames commented:
This is mainly waiting for #8 before I finalise first go at this

michaelwood commented on src/lib/web-rpde-query.js:
```diff
@@ -0,0 +1,81 @@
+  constructor(afterTimestamp, afterId, limit) {
+    this.afterTimestamp = parseInt(afterTimestamp) > 0 ? parseInt(afterTimestamp) : null;
+    this.afterId = afterId;
+    this.limit = parseInt(limit) > 0 ? parseInt(limit) : Settings.rpdeDefaultLimit;
```
Number(limit) would be fine here since we're not parsing something more complex such as `parseInt(34e3, 10)`  which is what I understand that function is really for.

odscjames replied:
ok

michaelwood commented on src/lib/web-rpde-query.js:
```diff
@@ -0,0 +1,81 @@
+    this.rows = []
+    this.dontServeDataUntilSecondsOld = Settings.rpdeDontServeDataUntilSecondsOld;
+  }
+
```
let's either switch to camelCase everywhere or use lower case and underscores

odscjames replied:
I'll change some but some specifically follow GET params set by the RPDE standard, so I'd like to keep those the same. Also, we seem to have established camelcase convention in settings so lets leave that and fix elsewhere if we want.

michaelwood commented on src/lib/web-rpde-query.js:
```diff
@@ -0,0 +1,81 @@
+
+
+
+class rpde_query {
```
Class name should be capitalised at least and probably no underscore

odscjames replied:
ok

michaelwood commented on src/lib/web-rpde-query.js:
```diff
@@ -0,0 +1,81 @@
+    }
+    // Finally order and limit
+    sql += "ORDER BY extract(epoch from updated_at)::int ASC, data_id ASC LIMIT " + this.limit
+    // ----- Run Query, store results
```
is this.limit safe? it looks like it could come through the http get query parameters. Using the $ placeholders like you've done elsewhere looks like the right thing to do

michaelwood replied:
I guess it comes through parseInt and should at least be a number, though you could upset parseInt .

odscjames replied:
Yes, because in the constructor it is def turned into a Int. But placeholder probably good for another safety thing.

michaelwood commented on src/lib/web-server.js:
```diff
@@ -9,6 +10,17 @@ web_server_app.get('/', (req, res) => {
+    res.json(
+        {
+            "next": "/normalised_data/all?" + query.get_next_get_params_string(),
+            "items": query.get_data()
```
this could get into a bit of race condition, couldn't run just return the data?

odscjames replied:
How? All processing is done in the run() function with one SQL statement. The get_next_get_params_string and get_data only return data stored on the object, and don't do any DB work of their own.

jarofgreen committed:
rpde: Changes requested in P.R.

https://github.com/openactive/conformance-services/pull/7

michaelwood commented on README.md:
```diff
@@ -75,6 +75,10 @@ To run this:
 
 `$ node ./src/bin/web-server.js`
 
+RPDE APIs are meant to be called directly by users. They are:
```
Shall we add a web server endpoints section here, then I can also add mine too

odscjames replied:
Because ours are different things, that's why I did diff sections. So far there is no user story for users accessing your directly, only the status page needs to. So I was thinking a different section with a note "There are meant to be accessed directly by the Status app, and are tailored to it" or something.

odscjames replied:
Basically, cos the status app is currently going to have no local caching the results of this end point will need to be tailored heavily to it and so we want to be able to change in regularly as we need to for the status page. So I don't want users using the same end point cos they will need stability. If a user story does emerge about users wanting to directly access it I'd rather they had diff end points.

But that's almost a side point. I don't really mind about this, it'll might get tweaked later when we put in different system anyway.

Do you want to merge this and rewrite in yours?

jarofgreen committed:
rpde: Add get_api_response for code readability, to avoid ordering errors and race conditions

https://github.com/openactive/conformance-services/pull/7

(Also comment on variable name schemes)
